src/main/java/de/tum/in/www1/artemis/service/scheduled/ProgrammingExerciseScheduleService.java

```diff
@@ -695,18 +696,21 @@ public Runnable runUnlockOperation(ProgrammingExercise exercise, BiConsumer<Prog
                     programmingExercise = programmingExerciseRepository.findByIdWithTemplateAndSolutionParticipationElseThrow(programmingExercise.getId());
                     unlockOperation.accept(programmingExercise, participation);
                 };
-                List<ProgrammingExerciseStudentParticipation> failedUnlockOperations = invokeOperationOnAllParticipationsThatSatisfy(programmingExerciseId,
-                        unlockAndCollectOperation, condition, "add write permissions to all student repositories");
-
-                // We send a notification to the instructor about the success of the repository unlocking operation.
-                long numberOfFailedUnlockOperations = failedUnlockOperations.size();
-                if (numberOfFailedUnlockOperations > 0) {
-                    groupNotificationService.notifyEditorAndInstructorGroupAboutExerciseUpdate(exercise,
-                            Constants.PROGRAMMING_EXERCISE_FAILED_UNLOCK_OPERATIONS_NOTIFICATION + failedUnlockOperations.size());
-                }
-                else {
-                    groupNotificationService.notifyEditorAndInstructorGroupAboutExerciseUpdate(exercise, Constants.PROGRAMMING_EXERCISE_SUCCESSFUL_UNLOCK_OPERATION_NOTIFICATION);
-                }
+                var failedUnlockOperationsFuture = invokeOperationOnAllParticipationsThatSatisfy(programmingExerciseId, unlockAndCollectOperation, condition,
+                        "add write permissions to all student repositories");
+
+                failedUnlockOperationsFuture.thenAccept(failedUnlockOperations -> {
+                    // We send a notification to the instructor about the success of the repository unlocking operation.
+                    long numberOfFailedUnlockOperations = failedUnlockOperations.size();
+                    if (numberOfFailedUnlockOperations > 0) {
+                        groupNotificationService.notifyEditorAndInstructorGroupAboutExerciseUpdate(exercise,
+                                Constants.PROGRAMMING_EXERCISE_FAILED_UNLOCK_OPERATIONS_NOTIFICATION + failedUnlockOperations.size());
+                    }
+                    else {
+                        groupNotificationService.notifyEditorAndInstructorGroupAboutExerciseUpdate(exercise,
+                                Constants.PROGRAMMING_EXERCISE_SUCCESSFUL_UNLOCK_OPERATION_NOTIFICATION);
+                    }
+                });
```

Comment:
Same two suggestions as above (rename `failedUnlockOperationsFuture` and use `notificationText` as temp variable for readability)